### PR TITLE
Much more extensive default sensitivity choices

### PR DESCRIPTION
**Pull request by ChrisRackauckas**
Fixes https://github.com/SciML/DiffEqSensitivity.jl/issues/246

### CODE REVIEW

File: src/adjoint_common.jl [hunk 1]
```diff
@@ -165,8 +165,20 @@ function adjointdiffcache(g,sensealg,discrete,sol,dg,f;quad=false,noiseterm=fals
         end
       end
     end
+
     if compile_tape(sensealg.autojacvec)
       paramjac_config = ReverseDiff.compile(tape)
+    elseif sensealg.autojacvec isa Bool && sensealg.autojacvec && DiffEqBase.isinplace(prob)
+      compile = try
+          !hasbranching(prob.f,copy(u0),u0,p,prob.tspan[2]))
```
Comment (ChrisRackauckas) [inline src/adjoint_common.jl:173]: @simeonschaub `prob.f` mutates the first argument: is the copy required here for safety?

File: src/concrete_solve.jl [hunk 2]
```diff
@@ -2,24 +2,41 @@
...
-      else
-        InterpolatingAdjoint()
-      end
+      InterpolatingAdjoint(autojacvec=ZygoteVJP())
+    end
+  else
+    # Determine if we can compile ReverseDiff
+    compile = try
+        if DiffEqBase.isinplace(prob)
+          !hasbranching(prob.f,copy(u0),u0,p,prob.tspan[1])
```
Comment (ChrisRackauckas) [inline src/concrete_solve.jl:26]: note copy.

File: src/adjoint_common.jl [hunk 3]
```diff
@@ -224,6 +236,17 @@ function adjointdiffcache(g,sensealg,discrete,sol,dg,f;quad=false,noiseterm=fals
           tapei = noisetape(i)
           if compile_tape(sensealg.noise)
             push!(paramjac_noise_config, ReverseDiff.compile(tapei))
+          elseif sensealg.noise isa Bool && sensealg.noise && DiffEqBase.isinplace(prob)
+              compile = try
+                  !hasbranching(prob.f,copy(u0),u0,p,prob.tspan[2]))
```
Comment (ChrisRackauckas) [inline src/adjoint_common.jl:241]: note copy

File: src/adjoint_common.jl (hunk 1 above)
Comment (simeonschaub) [reply src/adjoint_common.jl:173]: Yes, the overdub will always run everything as well, so this is needed if you don't want the check to mutate `u0`. Might even need to be a `deepcopy` if elements of `u0` are allowed to be mutated as well.

**Pull request merged**